### PR TITLE
Sum of evens

### LINKED ISSUE #38: Sum_Of_Evens Challenge
# Sum of Evens Challenge

## Description
Calculates the sum of all even numbers from 1 to a given number.

---

## Behavior
The function calculates the sum of all even integers within the inclusive range `[1, n]`.

## Examples

Example 1 
Input - sum_of_evens(10)
Output - 30 

Example 2
Input - sum_of_evens(5)
Output - 6

Example 3
Input - sum_of_evens(1)
Output - 0

### PR DESCRIPTION
**Pull request by marshar007007**
---
Sums of even numbers challenge
about: A template PR for code review with a checklist
---

### Files

- [x] The file name describes the function's behavior
- [x] There is a module docstring in the function file
- [x] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [x] There is a module docstring in the tests file

### Unit Tests

- [x] The test class has a helpful name in PascalCase
- [x] The test class has a docstring
- Every unit test has
  - [x] A helpful name
  - [x] A clear docstring
  - [x] Only one assertion
  - [x] There is no logic in the unit test
- [x] All tests pass
- [x] There are tests for defensive assertions
- [x] There are tests for boundary cases

### Function Docstring

- [x] The function's behavior is described
- The function's arguments are described:
  - [x] Type
  - [x] Purpose
  - [x] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [x] Type
  - [x] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [x] Each assumption about an argument is checked with an assertion
  - [x] Each assertion checks for _only one_ assumption about the argument
- [x] Include 3 or more (passing!) doctests

### The Function

- [x] The function's name describes it's behavior
- [x] The function's name matches the file name
  - _It's ok to have extra helper functions if necessary, like with mergesort_
- [x] The function has correct type annotations
- [x] The function is not called at the top level of the function file 
  - _Recursive solutions **can** call the function from **inside** the function body_

### Do's

- [x] Variable names help to understand the strategy
- [x] Any comments are clear and describe the strategy
- [x] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [x] The function's strategy _is not_ described in any docstrings or tests
- [x] Comments explain the _strategy_, **not** the _implementation_
- [x] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [x] The code passes the formatting checks
- [x] The code passes all Ruff linting checks
- [x] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [x] Variables are named with snake_case
- [x] Variable names are clear and helpful
- [x] The code follows the strategy as simply as possible
- [x] The implementation is as simple as possible given the strategy
- [x] There are no commented lines of code
- [x] There are no `print` statements anywhere
- [x] The code includes defensive assertions
- [x] Defensive assertions include as little logic as possible

### CODE REVIEW

File: solutions/tests/test_sum_of_evens [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+Test Categories:
+  - Regular Passing Cases: Valid ranges of inputs.
+  - Boundary Cases: Edge scenarios like zero or single values.
+  - Defensive Cases: Invalid or negative inputs.
+"""
+
+import unittest
+
+from sum_of_even import sum_of_evens
```
Comment (1sgtpepper) [inline solutions/tests/test_sum_of_evens]: The statement does not match the file name, which will cause an **`ImportError`** when running the tests.

File: solutions/sum_of_evens [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+        >>> sum_of_evens(1)
+        0
+    """
+    # Defensive assertions
+    assert isinstance(n, int), "Input must be an integer."
+    if n < 0:
+        raise ValueError("Input must be a non-negative integer.")
+
+    # Calculate sum of evens
+    return sum(num for num in range(1, n + 1) if num % 2 == 0)
```
Comment (1sgtpepper) [inline solutions/sum_of_evens]: You may consider optimizing the calculation using the arithmetic series formula` (count = n // 2; return count * (count + 1))` since it avoids iterating through the entire range, which can be inefficient for large inputs, but the current implementation is already good.

**Review by Kuhai9801 — DISMISSED** [review]
Excellent Work! The only major issue is the mistyped statement. Keep up the good work!

File: solutions/tests/test_sum_of_evens (hunk 1 above)
Comment (marshar007007) [reply solutions/tests/test_sum_of_evens]: thank u for your warm suggestions! I will make a change to it.

File: solutions/sum_of_evens (hunk 2 above)
Comment (marshar007007) [reply solutions/sum_of_evens]: if my implementation is good enough, I will leave this as original.

**Comment by 1sgtpepper** [comment]
Good work but fails py_formatting.

**Review by Kuhai9801 — APPROVED** [review]

**Comment by 1sgtpepper** [comment]
Good Work! Approved.
